include/nbl/builtin/hlsl/bxdf/fresnel.hlsl

```diff
@@ -17,131 +20,449 @@ namespace hlsl
+    }
+
+    Refract<scalar_type> refract;
+};
+
+
+namespace fresnel
+{
+
+#define NBL_CONCEPT_NAME Fresnel
+#define NBL_CONCEPT_TPLT_PRM_KINDS (typename)
+#define NBL_CONCEPT_TPLT_PRM_NAMES (T)
+#define NBL_CONCEPT_PARAM_0 (fresnel, T)
+NBL_CONCEPT_BEGIN(1)
+#define fresnel NBL_CONCEPT_PARAM_T NBL_CONCEPT_PARAM_0
+NBL_CONCEPT_END(
+    ((NBL_CONCEPT_REQ_TYPE)(T::scalar_type))
+    ((NBL_CONCEPT_REQ_TYPE)(T::vector_type))
+    ((NBL_CONCEPT_REQ_EXPR_RET_TYPE)((fresnel()), ::nbl::hlsl::is_same_v, typename T::vector_type))
+);
+#undef fresnel
+#include <nbl/builtin/hlsl/concepts/__end.hlsl>
+
+template<typename T NBL_PRIMARY_REQUIRES(concepts::FloatingPointLikeVectorial<T>)
+struct Schlick
+{
+    using scalar_type = typename vector_traits<T>::scalar_type;
+    using vector_type = T;
+
+    static Schlick<T> create(NBL_CONST_REF_ARG(T) F0, scalar_type clampedCosTheta)
+    {
+        Schlick<T> retval;
+        retval.F0 = F0;
+        retval.clampedCosTheta = clampedCosTheta;
+        return retval;
+    }
+
+    T operator()()
+    {
+        assert(clampedCosTheta > scalar_type(0.0));
+        assert(hlsl::all(hlsl::promote<T>(0.02) < F0 && F0 <= hlsl::promote<T>(1.0)));
+        T x = 1.0 - clampedCosTheta;
+        return F0 + (1.0 - F0) * x*x*x*x*x;
+    }
+
+    T F0;
+    scalar_type clampedCosTheta;
+};
+
+template<typename T NBL_PRIMARY_REQUIRES(concepts::FloatingPointLikeVectorial<T>)
+struct Conductor
+{
+    using scalar_type = typename vector_traits<T>::scalar_type;
+    using vector_type = T;
+
+    static Conductor<T> create(NBL_CONST_REF_ARG(T) eta, NBL_CONST_REF_ARG(T) etak, scalar_type clampedCosTheta)
+    {
+        Conductor<T> retval;
+        retval.eta = eta;
+        retval.etak = etak;
+        retval.etak2 = etak*etak;
+        retval.clampedCosTheta = clampedCosTheta;
+        return retval;
+    }
+
+    static Conductor<T> create(NBL_CONST_REF_ARG(complex_t<T>) eta, scalar_type clampedCosTheta)
+    {
+        Conductor<T> retval;
+        retval.eta = eta.real();
+        retval.etak = eta.imag();
+        retval.etak2 = eta.imag()*eta.imag();
+        retval.clampedCosTheta = clampedCosTheta;
+        return retval;
+    }
+
+    T operator()()
+    {
+        const scalar_type cosTheta2 = clampedCosTheta * clampedCosTheta;
+        //const float sinTheta2 = 1.0 - cosTheta2;
+
+        const T etaLen2 = eta * eta + etak2;
+        assert(hlsl::all(etaLen2 > hlsl::promote<T>(hlsl::exp2<scalar_type>(-numeric_limits<scalar_type>::digits))));
+        const T etaCosTwice = eta * clampedCosTheta * 2.0f;
+
+        const T rs_common = etaLen2 + (T)(cosTheta2);
+        const T rs2 = (rs_common - etaCosTwice) / (rs_common + etaCosTwice);
+
+        const T rp_common = etaLen2 * cosTheta2 + (T)(1.0);
+        const T rp2 = (rp_common - etaCosTwice) / (rp_common + etaCosTwice);
+
+        return (rs2 + rp2) * 0.5f;
+    }
+
+    T eta;
+    T etak;
+    T etak2;
```

Comment:
it seems the thing you really want to precompute would be `etaLen2` 

and store `eta`

so 2 not 3 variables